GHzDACs/jumpTable.py

```diff
@@ -4,182 +4,234 @@
 import numpy as np
 from util import littleEndian
 
-JUMP_INDEX_MIN = 1
+JUMP_INDEX_MIN = 0
 JUMP_INDEX_MAX = 15
 IDLE_MIN_CYCLES = 1
-IDLE_MAX_CYCLES = (2**15)-1
+IDLE_MAX_CYCLES = (2 ** 15) - 1
 DAISY_VALUE_MIN = 0
 DAISY_VALUE_MAX = 15
 
 NUM_COUNTERS = 4
 
 SRAM_ADDR_MIN = 0
-SRAM_ADDR_MAX = 8192 - 1 #XXX HACK ALERT! This should be imported from other module or registry!!!
+SRAM_ADDR_MAX = 8192 - 1  # XXX HACK ALERT! This should be imported from other module or registry!!!
 
 # A single entry in the jump table
 
+
 class JumpEntry(object):
-    """A single entry in the jump table"""
-    def __init__(self, fromAddr, toAddr, operation=None):
-        self.fromAddr = fromAddr
-        self.toAddr = toAddr
+    """ A single entry in the jump table.
+
+    * Every from address has 2 subtracted to account for FPGA timing offsets.
+    * The END entry has an additional 1 subtracted.
+    * IDLE has one clock cycle subtracted from idle length as well.
+
+    A note about from addresses:
+
+    We subtract 2 from the from_addr passed into the JumpEntry. The reason for
+    this is that by convention we set the end address of an algorithm to the
+    length of the sequence. That is, if the SRAM is length 256 ns, then we set
+    the end address to 256//4 == 64. However, if we want to idle over the last
+    4 ns/jump after the last 4 ns/whatever, then we need to notice that those
+    4ns are from 252 - 255, i.e. address 63. So we set the from_address to 62,
+    because the actual fromAddress must be one less than the address you want,
+    due to FPGA offsets, per John's documentation. (In the case of END, it's
+    two less, and that's handled as well.)
+    """
+
+    def __init__(self, from_addr, to_addr, operation):
+        """ Create a single jump table entry.
+
+        :param int from_addr: the from address
+        :param int to_addr: the to address (should be 0 if not used)
+        :param Operation operation: the operation
+        """
+        if abs(from_addr - to_addr) - 1 == 0:
+            raise ValueError(
+                "from_addr: {} and to_addr: {} are too close".format(
+                    from_addr, to_addr)
+            )
+        self._to_addr = 0
+        self._from_addr = 0
+        self.from_addr = from_addr - 2  # account for offsets, see docstring
+        self.to_addr = to_addr
         self.operation = operation
-    
-    def getToAddr(self):
-        return self._toAddr
-    def setToAddr(self, addr):
-        if SRAM_ADDR_MIN <= addr and addr <= SRAM_ADDR_MAX:
-            self._toAddr = addr
+
+        if self.operation.NAME == 'END':
+            self.from_addr -= 1  # account for additional END offset
```

Comment:
I'm not sure think these jump table objects are the right place to deal with offsets like this. It seems to me that these should just be data classes, and the logic of what addresses and parameters to choose should be contained in the code that _constructs_ the jump table, not in the jump table classes themselves. In particular, these leads to strangeness like:
```python
e = JumpEntry(from_addr=2, to_addr=10, operation=IDLE(10))
assert(e.from_addr == 2) # fails
e.from_addr = 2
assert(e.from_addr == 2) # succeeds
```
This brings up another point, which is that all of these classes should probably be immutable. That would argue for doing the range checking in the constructor and not providing setters for these parameters at all, e.g. using a read-only property.